src/Streamly/Internal/Data/Array/Stream/Foreign.hs

```diff
@@ -457,18 +525,23 @@ _toArraysOf n = FL.chunksOf n (A.writeNF n) FL.toStream
 -}
 
 -------------------------------------------------------------------------------
--- Parsing
+-- Elimination - running element parsers
 -------------------------------------------------------------------------------
 
 -- GHC parser does not accept {-# ANN type [] NoSpecConstr #-}, so we need
 -- to make a newtype.
 {-# ANN type List NoSpecConstr #-}
 newtype List a = List {getList :: [a]}
 
+-- This can be generalized to any type provided it can be unfolded to a stream
+-- and it can be combined using a semigroup operation.
+--
+-- XXX This should be written using CPS (as parseK) if we want it to scale wrt
+-- to the number of times it can be called on the same stream.
 {-# INLINE_NORMAL parseD #-}
 parseD ::
        forall m a b. (MonadIO m, MonadThrow m, Storable a)
-    => PRD.Parser m (Array a) b
+    => PRD.Parser m a b
```

Comment:
Can you compare the performance of this with `parseArray` in https://github.com/composewell/streamly/pull/912/

Reply:
I'm guessing the one in #912 would be more efficient if there is a lot of backtracking.

Reply:
> Can you compare the performance of this with `parseArray` in #912

I will let you do that. If you can create a benchmark that proves that the DLL implementation is significantly better we will replace it with that.